### PR TITLE
[Model] Support official SenseNova-U1 LoRA checkpoints

### LINKED ISSUE #4705: [New Model]: SenseNova-U1-8B-MoT-LoRAs
### The model to consider.

https://huggingface.co/sensenova/SenseNova-U1-8B-MoT-LoRAs

### The closest model vllm-omni already supports.

https://huggingface.co/sensenova/SenseNova-U1-8B-MoT

### What's your difficulty of supporting the model you want?

vLLM-Omni already supports the SenseNova-U1-8B-MoT model series. Does it also support SenseNova-U1-8B-MoT LoRAs? If yes, how can single-file SenseNova-U1 LoRAs (.safetensors) be converted into standard PEFT adapter directories compatible with vLLM-Omni? Could you provide corresponding technical support?

### Use case and motivation

The SenseNova-U1-8B-MoT model was trained on 2K resolution images. It produces severely degraded, practically unusable outputs at other resolutions. However, when deployed via vLLM-Omni, generating 2K images with this model takes over 50 seconds. Therefore, we urgently need to utilize the SenseNova-U1-8B-MoT-LoRA-8step-V1.0.safetensors file to boost generation performance.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by EchoHayate**
## Purpose

Closes #4705.

The released `sensenova/SenseNova-U1-8B-MoT-LoRAs` checkpoints use a
single-file layout with `lora_down`, `lora_up`, and per-layer `alpha` tensors,
while vLLM-Omni's diffusion LoRA runtime expects PEFT format. They also include
adapters for `fm_modules.fm_head`, which used native `nn.Linear` layers and
therefore was not visible to the existing vLLM LoRA wrappers.

This PR:

- adds `tools/sensenova_u1/convert_lora_to_peft.py` to convert an official
  checkpoint into `adapter_model.safetensors` and `adapter_config.json`;
- rejects malformed pairs, inconsistent rank/alpha, and zero-rank adapters
  before writing output;
- changes the two non-pixel `fm_head` projections to replicated vLLM linear
  layers without changing their checkpoint parameter names;
- exposes `fm_modules` to `DiffusionLoRAManager`, enabling request-scoped
  activation and deactivation of the official `fm_head` adapters;
- documents the conversion workflow and adds converter, base-weight loading,
  and real manager round-trip tests.

The implementation is based on current `main` and preserves the typed
SenseNova configuration and request-local `t_eps` ownership introduced by
#5877.

## Test Plan

```bash
python -m pytest -q \
  tests/tools/test_convert_sensenova_u1_lora.py \
  tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py \
  tests/diffusion/lora/test_lora_manager.py

pre-commit run --files \
  docs/user_guide/diffusion/lora.md \
  tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py \
  tests/tools/test_convert_sensenova_u1_lora.py \
  tools/sensenova_u1/convert_lora_to_peft.py \
  vllm_omni/diffusion/models/sensenova_u1/pipeline_sensenova_u1.py
```

## Current-head validation

**Current base:** `b78c31eda6c75546be4d54c265d483977104a487`

**Current PR head:** `c326f201bbd79747804bcb8436eac3268f0b0002`

- Refreshed onto current `main`; the only textual conflict was in
  `docs/user_guide/diffusion/lora.md`. The resolution preserves both this PR's
  SenseNova conversion section and upstream's MiniMax-H3 adapter-schedule
  documentation.
- The net PR diff remains limited to the same five SenseNova LoRA files.
- Converter tests in an isolated Python 3.12 environment:
  `13 passed`.
- Changed-file pre-commit: all applicable hooks passed, including Ruff,
  mypy, pytest marker policy, Markdown, SPDX, and forbidden-import checks.
- Python 3.12 byte compilation and `git diff --check`: passed.
- The full model/manager CPU suite and A100 E2E have not yet been rerun at
  `c326f201`; GitHub base CI is running.

## Previously validated behavior

At head `afc530532c783f17697c23775ef1004b4dd9b138`:

- converter + SenseNova runtime + generic LoRA manager:
  `34 passed, 15 warnings`;
- all existing `fm_head.{0,2}.{weight,bias}` checkpoint names loaded through
  `SenseNovaU1Pipeline.load_weights()`;
- the real `DiffusionLoRAManager` wrapped and loaded both `fm_head.0` and
  `fm_head.2`; activation changed output and deactivation restored the exact
  baseline.

These results are retained as historical functional evidence, not as
current-head verification.

### Historical official-adapter validation

- Source:
  `SenseNova-U1-8B-MoT-LoRA-8step-V1.0.safetensors`
- Source size: 1,637,476,232 bytes
- Source SHA-256:
  `3fc8c94330ee16c210b294189ebf3b88f41693f2ddfc5db21239c8cc270024e0`
- Converted adapter SHA-256:
  `c8aecd009cfa1696c18fbe495a62a77161d56922791ec10884f7f96638a8a307`
- Converted tensors: 592
- Target modules: 296
- Rank / alpha: 128 / 128
- Includes `fm_modules.fm_head.0` and `fm_modules.fm_head.2`

### Historical full image-generation E2E (A100 80GB)

The E2E was produced at the previous exact base/head pair:

- exact base:
  `73b623f2f7db092053c1c86fe796bed89eb3dc71`
- exact head:
  `afc530532c783f17697c23775ef1004b4dd9b138`
- model: `sensenova/SenseNova-U1-8B-MoT`
- model revision:
  `bfa9b436503cb8aed4f2bc60e3236710cc77468d`
- prompt: `a photo of a cat sitting on a laptop keyboard`
- sampling: seed 42, 512x512, 8 steps, batch size 1, CFG scale 4.0

| Comparison | Result |
|---|---:|
| Exact base baseline vs exact head baseline | byte-identical, pixel MAE 0.0 |
| Head baseline vs official LoRA | different, pixel MAE 97.256365, max abs diff 255 |
| Head baseline vs post-deactivation request | byte-identical, pixel MAE 0.0 |

All four outputs were finite, non-constant RGB 512x512 images. The head run
reused one production `Omni` engine for baseline, official-LoRA activation,
and deactivation.

No compatibility overlay or source patch was applied to either checkout.
This is correctness validation, not a throughput claim. The first LoRA request
includes adapter loading, and the base/head initialization timings were
affected by cold versus warm filesystem cache.

### CODE REVIEW

**Comment by EchoHayate** [comment]
## Full SenseNova-U1 + official LoRA E2E validation

I completed the previously missing full image-generation validation on an NVIDIA A100 80GB using the production `Omni` path (not a converter-only or mocked harness).

### Fixed inputs

- Base commit: `0a51a18570d8df8e65e9d4f9be688ce686c67ce6`
- PR head: `62690e2b9e1d4d943601aa430e5648d753f18a7d`
- Base model: `sensenova/SenseNova-U1-8B-MoT`
- Model revision: `bfa9b436503cb8aed4f2bc60e3236710cc77468d`
- Model validation: 16/16 files checked, 35,110,117,492 bytes total; all eight safetensors shards matched their Hub SHA-256 metadata
- Official source LoRA: `SenseNova-U1-8B-MoT-LoRA-8step-V1.0.safetensors`
  - 1,637,476,232 bytes
  - SHA-256 `3fc8c94330ee16c210b294189ebf3b88f41693f2ddfc5db21239c8cc270024e0`
- Converted PEFT adapter:
  - 1,637,446,080 bytes
  - SHA-256 `c8aecd009cfa1696c18fbe495a62a77161d56922791ec10884f7f96638a8a307`
  - rank/alpha 128/128, 296 target modules
- Prompt: `a photo of a cat sitting on a laptop keyboard`
- Sampling: seed 42, 512x512, 8 steps, batch size 1, CFG scale 4.0

### Correctness results

| Comparison | Result |
|---|---:|
| Base commit baseline vs PR-head baseline | byte-identical, pixel MAE 0.0 |
| PR-head baseline vs official LoRA | different, pixel MAE 97.256365, max abs diff 255 |
| PR-head baseline vs request after LoRA deactivation | byte-identical, pixel MAE 0.0 |

All generated PNGs were finite, non-constant RGB images. This verifies the complete sequence through the real diffusion engine: initialize SenseNova-U1, generate without an adapter, load and activate all 296 official adapter modules for one request, then deactivate the adapter and exactly restore baseline behavior.

### Timing and memory

| Run | Init | Request wall time | Peak GPU memory |
|---|---:|---:|---:|
| Base baseline | 26.141 s | 1.591 s | 47,383 MiB |
| PR-head baseline | 33.179 s | 1.901 s | 48,041 MiB |
| PR-head first official-LoRA request | same engine | 22.619 s | included above |
| PR-head request after deactivation | same engine | 1.329 s | included above |

The first LoRA request includes adapter loading/conversion work and is **not** presented as steady-state throughput.

### Compatibility boundary

The validation image uses vLLM 0.26, which requires `_DiffusionVllmModelConfig.head_dtype`. The one-property compatibility fix from #5766 (`fbee7d3ac437387e957b33914342d152186ba24b`, `head_dtype -> dtype`) was applied identically to both exact base and PR-head source snapshots. No #5642-specific code was changed in the base snapshot. Therefore, the A/B result isolates this PR while explicitly depending on that independent compatibility fix.

**Comment by EchoHayate** [comment]
Updated this PR onto current `main` and force-updated the existing branch to
`afc530532c783f17697c23775ef1004b4dd9b138`.

Key changes since the previous head:

- preserves the typed SenseNova config and request-local `t_eps` ownership from
  #5877;
- adds malformed-input and zero-rank validation to the official converter;
- exercises the real `DiffusionLoRAManager` activation/deactivation path;
- removes the old #5766 compatibility-overlay dependency.

Fresh verification:

- converter + SenseNova runtime + generic manager:
  `34 passed, 15 warnings`;
- changed-file pre-commit: all applicable hooks passed;
- exact A100 80GB E2E, base `73b623f2` vs head `afc53053`:
  - main baseline == head baseline byte-for-byte;
  - official LoRA changes the image (pixel MAE `97.256365`);
  - deactivation restores the baseline byte-for-byte.

Both E2E source trees were clean Git checkouts, imported directly from their
mounted source roots, and used the pinned model revision
`bfa9b436503cb8aed4f2bc60e3236710cc77468d`. No compatibility overlay or source
patch was applied. The PR body now contains the full current-main validation
contract and boundaries.

Fresh base CI is running. Review is welcome once it is green.

**Comment by EchoHayate** [comment]
## Current-head refresh / author self-review

- Refreshed onto `main@b78c31ed`; current head is `c326f201`. The only textual conflict was `docs/user_guide/diffusion/lora.md`, resolved by preserving both this PR's SenseNova conversion instructions and upstream's MiniMax-H3 adapter-schedule section.
- Rechecked the net diff: it remains limited to five SenseNova LoRA files and still closes one loop—official checkpoint conversion, `fm_modules` visibility to `DiffusionLoRAManager`, request-scoped activation/deactivation, and documentation.
- Base checkpoint names remain unchanged; malformed pairs, inconsistent rank/alpha, zero-rank tensors, unknown tensors, and unsafe overwrite remain fail-closed.
- Fresh current-head local evidence: converter suite `13 passed`; changed-file pre-commit, Ruff check/format, Python 3.12 byte compilation, and `git diff --check` passed.
- The prior `34 passed` model/manager suite and A100 image-generation A/B were produced at `afc53053` and are now explicitly labeled historical in the PR body. They are not presented as current-head GPU evidence. GitHub base CI is running for `c326f201`.

@xuechendi @SamitHuang, this refreshed PR is ready for review once the current checks finish.

Current head reviewed: `c326f201bbd79747804bcb8436eac3268f0b0002`.

File: tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    output = fm_head(torch.randn(2, 3, 8))
+    assert output.shape == (2, 3, 12)
+
+
+def test_sensenova_lora_manager_scans_fm_modules():
+    assert SenseNovaU1Pipeline._lora_components == ["language_model", "fm_modules"]
+
+
+def test_sensenova_fm_head_preserves_checkpoint_weight_names():
+    pipeline = torch.nn.Module()
```
Comment (lishunyang12) [inline tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py]: Build this stub as SenseNovaU1Pipeline so stacked_params_mapping is available.

File: tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    )
+
+    os.environ.setdefault("MASTER_ADDR", "127.0.0.1")
+    os.environ["MASTER_PORT"] = str(get_open_port())
+    with set_current_vllm_config(VllmConfig()):
+        init_distributed_environment(
+            world_size=1,
+            rank=0,
+            local_rank=0,
+            distributed_init_method="env://",
```
Comment (lishunyang12) [inline tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py:52]: Use backend="gloo" so these CPU tests do not require CUDA.

File: tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py (hunk 1 above)
Comment (EchoHayate) [reply tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py]: Fixed in 9419599d. The weight-name test now allocates a SenseNovaU1Pipeline without loading a checkpoint, initializes its nn.Module state, and calls the bound production load_weights method. It inherits stacked_params_mapping instead of using a plain Module. The original test reproduced the AttributeError before the change; all four tests in this file now pass, with the existing loaded-name and tensor-equality assertions unchanged.

File: tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py (hunk 2 above)
Comment (EchoHayate) [reply tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py:52]: Fixed in 9419599d by explicitly passing backend="gloo" to init_distributed_environment. Model-parallel initialization inherits the initialized world-group backend. The CPU suite passes locally (4 tests); the combined converter, SenseNova LoRA, distilled-LoRA, and manager regression set passes 44 tests with 15 warnings. No production backend default was changed.

**Comment by EchoHayate** [comment]
## Review follow-up / author self-review

Current head reviewed: `9419599d6560dbb914cc28efa367cc599b56a77b`.

- This update changes only `tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py`: an explicit Gloo backend for the CPU fixture, and a real `SenseNovaU1Pipeline` stub for the checkpoint-weight-name test. No production code, loading semantics, assertions, or tolerances changed.
- Reproduced the original missing-`stacked_params_mapping` failure before the fix. The corrected test calls the bound production loader and retains exact loaded-key and tensor-equality checks.
- Focused file: **4 passed**. Combined converter, SenseNova LoRA, distilled-LoRA and generic manager tests: **44 passed, 15 warnings**.
- Changed-file pre-commit passed, including mypy, test marks and SPDX; Ruff check/format and `git diff --check` passed.
- Tests ran on local macOS CPU with Python 3.12.13, PyTorch 2.11.0 and vLLM source `98dff2a81d747d1dba01a47f939f48c3526d4206`. These are current-head CPU results, not new GPU or full-checkpoint image-generation evidence. The older A100 results remain historical.

Reproduction command, with the PR checkout and the above vLLM source on `PYTHONPATH`, and `HF_HUB_OFFLINE=1`:

```bash
python -m pytest \
  tests/tools/test_convert_sensenova_u1_lora.py \
  tests/diffusion/models/sensenova_u1/test_sensenova_u1_lora.py \
  tests/diffusion/models/sensenova_u1/test_sensenova_u1_distill_lora.py \
  tests/diffusion/lora/test_lora_manager.py \
  -o addopts='' -q
```
